src/pubsub.c

```diff
@@ -615,6 +615,116 @@ void publishCommand(client *c) {
     addReplyLongLong(c,receivers);
 }
 
+typedef struct {
+    dict *subscription_dict;
+    long count;
+    int use_pattern;
+    sds pat;
+} pubsubSubscribersCmdArgs;
+
+static inline void pubsubSubscribersCmdArgsInitializeDefault(pubsubSubscribersCmdArgs *args) {
+    args->count = 1000;
+    args->use_pattern = 0;
+}
+
+static void pubsubReplySubscribersClientInfo(client *c, list *clients) {
+    listNode *ln;
+    listIter li;
+    listRewind(clients, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        client *client = listNodeValue(ln);
+        addReplyMapLen(c, 3);
+        addReplyBulkCString(c, "name");
+        if (client->name) {
+            addReplyBulk(c, client->name);
+        } else {
+            addReplyNull(c);
+        }
+        addReplyBulkCString(c, "id");
+        addReplyBulkLongLong(c, client->id);
+        addReplyBulkCString(c, "addr");
+        addReplyBulkCString(c, getClientPeerId(client));
+    }
+}
+
+/* Parse the arguments for PUBSUB SUBSCRIBERS command */
+static int parseArgsPubSubSubscribersCmd(client *c, pubsubSubscribersCmdArgs *args) {
+    if (c->argc < 3) {
+        addReplySubcommandSyntaxError(c);
+        return C_ERR;
+    }
+
+    /* Mandatory parameter - subscription_type */
+    char *type = c->argv[2]->ptr;
+    if (!strcasecmp(type, "channel")) {
+        args->subscription_dict = server.pubsub_channels;
+    } else if (!strcasecmp(type, "pattern")) {
+        args->subscription_dict = server.pubsub_patterns;
+    } else if (!strcasecmp(type, "shard-channel")) {
+        args->subscription_dict = server.pubsubshard_channels;
+    } else {
+        addReplySubcommandSyntaxError(c);
+        return C_ERR;
+    }
+
+    /* Optional parameter - match / count */
+    for (int i = 3; i < c->argc; i++) {
+        int moreargs = c->argc-i-1;
+        char *o = c->argv[i]->ptr;
+        if (!strcasecmp(o, "match") && moreargs) {
+            i++;
+            args->pat = c->argv[i]->ptr;
+            args->use_pattern = !(sdslen(args->pat) == 1 && args->pat[0] == '*');
+        } else if (!strcasecmp(o, "count") && moreargs) {
+            i++;
+            if (getRangeLongFromObjectOrReply(c, c->argv[i], -1, LONG_MAX, &args->count,
+                                             "count should be greater than or equal to -1") != C_OK) {
+                return C_ERR;
+            }
+            /* count arg provided as -1 signifies all the channels needs to be sent. */
+            if (args->count == -1) {
+                args->count = dictSize(args->subscription_dict);
+            }
+        } else {
+            addReplySubcommandSyntaxError(c);
+            return C_ERR;
+        }
+    }
+    return C_OK;
+}
+
+static void pubsubReplySubscribers(client *c) {
+    /* Initialize default args and parse if any. */
+    pubsubSubscribersCmdArgs args;
+    pubsubSubscribersCmdArgsInitializeDefault(&args);
+    if (parseArgsPubSubSubscribersCmd(c, &args) != C_OK) {
+        return;
+    }
```

Comment:
A minor preference: I generally dislike parse functions and parsed argument structs unless they are re-used between multiple calls. I would rather the top level redis command do the parsing, and then call a subcommand with the arguments.

Reply:
The size of the parsing code is around 40 lines. So, keeping it separate still keeps the top level redis command function compact. 

I'm moving the invocation of parsing the arguments though to the top level redis command.

Reply:
Sure. Again, not a strong preference. I find the extra abstraction of having an args struct not worth trying to keep it to 40 lines, but I'm okay with it.